### PR TITLE
Stop using atexit in tests

### PR DESCRIPTION
**Pull request by murgatroid99**
Add `grpc_test_before_shutdown` function to use instead of `atexit`, and add `grpc_test_shutdown` function to trigger the scheduled functions.

Note: Calls to `grpc_test_shutdown` were added with a sed script, so it's possible that some shouldn't actually be there.

### CODE REVIEW

File: test/core/util/test_config.c [hunk 1]
```diff
@@ -281,4 +294,30 @@ void grpc_test_init(int argc, char **argv) {
...
+    shutdown_handlers.head = shutdown_handlers.tail = elem;
+  } else {
+    shutdown_handlers.tail->next = elem;
+    shutdown_handlers.tail = elem;
+  }
+}
+
+void grpc_test_shutdown() {
+  for (shutdown_handler_list_elem *handler = shutdown_handlers.head;
+       handler != NULL; handler = handler->next) {
```
Comment (nicolasnoble) [inline test/core/util/test_config.c]: That's a use after free. (handler = handler->next)

**Comment by ctiller** [comment]
Why is this desirable?

**Comment by murgatroid99** [comment]
We currently use `atexit` to clean up ports requested from the port server. Doing that touches C internals (`grpc_pollset_work`), so it should not be called after `grpc_shutdown`.

**Comment by ctiller** [comment]
But we bracket that with a grpc_init and grpc_shutdown pair already, so the
pollset_work is not called after grpc_shutdown.

**Comment by murgatroid99** [comment]
I see. I guess the problem I was facing only applies to libuv, because the uv implementation has fake pollsets. Every pollset polls on everything, so the pollset in `atexit` was polling on things that were cleaned up in the previous `grpc_shutdown`. I'm OK with closing this and fixing that back in the uv pull request if we don't expect this to be a problem anywhere else.

**Comment by ctiller** [comment]
I think that would be better: it's more correct (init/shutdown are forced
to do the right thing), and it's less of a tax on every test that we write
(and less taxes there is a really hugely good thing).

**Pull request closed without merging**